### PR TITLE
[Bifrost] Implements seal() for LocalLoglet

### PR DESCRIPTION
**Pull request by AhmedSoliman**
[Bifrost] Implements seal() for LocalLoglet

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1718).
* #1726
* #1725
* #1722
* #1720
* __->__ #1718
* #1717
* #1705
* #1704

### CODE REVIEW

File: crates/bifrost/src/providers/local_loglet/mod.rs [hunk 1]
```diff
@@ -230,14 +234,31 @@ impl LogletBase for LocalLoglet {
...
+        let release_pointer = LogletOffset::from(
+            self.last_committed_offset
+                .fetch_max(offset.into(), Ordering::SeqCst)
+                .max(offset.into()),
+        );
+        self.notify_readers(release_pointer);
+        // Ensure that we don't acknowledge the append (even that it has happened) if the loglet
+        // has been sealed already. Memory ordering Acquire is required to establish happens-before
+        // relationship between bumping the last_commit_offset and the seal operation.
+        if self.sealed.load(Ordering::Acquire) {
```
Comment (tillrohrmann) [inline crates/bifrost/src/providers/local_loglet/mod.rs:248]: I think I don't fully understand the reasoning here. If we read `sealed == true`, it makes sense to abort the append. Couldn't this read be done with `Ordering::Relaxed` since we don't depend on any other memory writes that happened before setting `sealed = true`? If we read `sealed == false`, this is probably the dangerous situation since we don't know whether the write (self.sealed = true) is still in the store buffer and we have already read `self.last_committed_offset` in `find_tail` before we incremented it in this method?

**Review by tillrohrmann — COMMENTED** [review]
Thanks a lot for implementing seal for the `LocalLoglet`! Whenever I have to deal with memory orderings, I am on thin ice. So maybe the comments I've left don't make sense at all. My main question was to clarify my understanding about why certain accesses are `SeqCst` and `Acquire`/`Release` and how the problematic scenario of someone having sealed the loglet and reported a tail and a concurrent append operation being prevented from succeeding works.

File: crates/bifrost/src/providers/local_loglet/mod.rs [hunk 2]
```diff
@@ -491,4 +534,87 @@ mod tests {
...
+                    LocalLoglet::create(
+                        params
+                            .id()
+                            .parse()
+                            .expect("loglet params can be converted into u64"),
+                        log_store,
+                        log_writer,
+                    )
+                    .await?,
+                );
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/bifrost/src/providers/local_loglet/mod.rs]: I can do that in a future follow up PR

File: crates/bifrost/src/providers/local_loglet/mod.rs [hunk 3]
```diff
@@ -230,14 +234,31 @@ impl LogletBase for LocalLoglet {
...
+                .fetch_max(offset.into(), Ordering::SeqCst)
+                .max(offset.into()),
+        );
+        self.notify_readers(release_pointer);
+        // Ensure that we don't acknowledge the append (even that it has happened) if the loglet
+        // has been sealed already. Memory ordering Acquire is required to establish happens-before
+        // relationship between bumping the last_commit_offset and the seal operation.
+        if self.sealed.load(Ordering::Acquire) {
+            return Err(AppendError::Sealed);
+        }
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/bifrost/src/providers/local_loglet/mod.rs]: I really like the idea of fetch_add(0, Release), very clever solution! 

In all fairness, my approach was to mash something together and slam SeqCst as temporary measure until a follow up PR that will replace the entire module with a sequencer implementation. But since you've done a great job reviewing it, I've updated this version to a nicer/better version. 

In fact, what was making the whole thing work is the Mutex that already acts as a memory barrier/fence which synchronized threads.

File: crates/bifrost/src/providers/local_loglet/mod.rs [hunk 4]
```diff
@@ -295,7 +329,7 @@ impl LogletBase for LocalLoglet {
     /// new_trim_point is inclusive (will be trimmed)
     async fn trim(&self, new_trim_point: Self::Offset) -> Result<(), OperationError> {
         let effective_trim_point = new_trim_point.min(LogletOffset(
-            self.last_committed_offset.load(Ordering::Relaxed),
+            self.last_committed_offset.load(Ordering::Acquire),
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/bifrost/src/providers/local_loglet/mod.rs]: Not needed

File: crates/bifrost/src/providers/local_loglet/mod.rs [hunk 5]
```diff
@@ -329,7 +363,16 @@ impl LogletBase for LocalLoglet {
...
-        todo!()
+        if self.sealed.load(Ordering::Acquire) {
+            return Ok(());
+        }
+        let receiver = self.log_writer.enqueue_seal(self.log_id).await?;
+        let _ = receiver.await.unwrap_or_else(|_| {
+            warn!("Unsure if the local loglet record was sealed, the ack channel was dropped");
+            Err(ShutdownError.into())
+        })?;
+        self.sealed.store(true, Ordering::Release);
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/bifrost/src/providers/local_loglet/mod.rs]: Not needed anymore.

**Comment by AhmedSoliman** [comment]
I've cleaned up and fixed this implementation, improved the test to tackle the corner case a little harder (find_tail during appends + seal), and validated the approach in an external loom test (albeit, loom tests can't catch all issues with Relaxed ordering, so I opted to use `AcqRel` in fetch_max which arguably can be relaxed more.

**Review by tillrohrmann — APPROVED** [review]
Great work @AhmedSoliman. As far as I can tell, the atomic operations seem sound to me. +1 for merging.

I've never used Loom so I'd be interested in taking a look at the test and to run it to see how it works :-)
